File: base_pkg/scripts/filter_topic.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#Authors : BEE
# kinematic.
# DATE: 21/06/2021
# AUTHOR: HOANG VAN QUANG

import rospy
import sys
import time
import threading
import signal

from geometry_msgs.msg import Twist
from message_pkg.msg import Driver_respond

from sti_msgs.msg import Velocities

from math import sin , cos , pi , atan2

class filterTopic():
	def __init__(self):
		print("ROS Initial!")
		rospy.init_node('filter_topic', anonymous=False) # False
		

		# -- parameter
		self.pub_frequence = rospy.get_param("frequence", 25.)
		self.topicSub_driver1 = rospy.get_param("topicSub_driver1", "/driver1_respond")
		self.topicSub_driver2 = rospy.get_param("topicSub_driver2", "/driver2_respond")
		self.topicPub_driver1 = rospy.get_param("topicPub_driver1", "/driver1_filter")
		self.topicPub_driver2 = rospy.get_param("topicPub_driver2", "/driver2_filter")

		self.rate = rospy.Rate(self.pub_frequence)
		# -- SUB
		rospy.Subscriber(self.topicSub_driver1, Driver_respond, self.driver1Respond_callback)
		self.driver1_respond = Driver_respond()

		rospy.Subscriber(self.topicSub_driver2, Driver_respond, self.driver2Respond_callback)
		self.driver2_respond = Driver_respond()

		# -- PUB
		self.pub_filter1 = rospy.Publisher(self.topicPub_driver1, Driver_respond, queue_size=50)
		self.driver1_filter = Driver_respond()

		self.pub_filter2 = rospy.Publisher(self.topicPub_driver2, Driver_respond, queue_size=50)
		self.driver2_filter = Driver_respond()
		# --
		self.time_pub = 1/self.pub_frequence # s
		self.preTime_pub1 = time.time()
		self.preTime_pub2 = time.time()
		# -- 
		self.nowTime_speed1 = time.time()
		self.nowTime_speed2 = time.time()
		self.is_timeout1 = 0
		self.is_timeout2 = 0
		self.timeout = self.time_pub*3.

	def driver1Respond_callback(self, data):
		self.driver1_respond = data
		self.nowTime_speed1 = time.time()

	def driver2Respond_callback(self, data):
		self.driver2_respond = data
		self.nowTime_speed2 = time.time()

	def run(self):
		print ("Launch ALL!")
		while not rospy.is_shutdown(): # not self.shutdown_flag.is_set() or 
			# --
			t1_out = time.time() - self.nowTime_speed1
			if (t1_out > self.timeout):
				self.is_timeout1 = 1
			else:
				self.is_timeout1 = 0
			# --
			t2_out = time.time() - self.nowTime_speed2
			if (t2_out > self.timeout):
				self.is_timeout2 = 1
			else:
				self.is_timeout2 = 0
			# --
			if (self.is_timeout1 == 0):
				self.pub_filter1.publish(self.driver1_respond)
			# --
			if (self.is_timeout2 == 0):
				self.pub_filter2.publish(self.driver2_respond)

			self.rate.sleep()
		print('program stopped')

def main():
	print('Starting main program')

	program = filterTopic()
	program.run()

	print('Exiting main program')	

if __name__ == '__main__':
    main()

File: cplus_pkg/basecplus_pkg/src/filter_topic.cpp
// Author : Phùng Quý Dương 
// Date: 21-09-2023

/*
    - Node filter_topic
    - function:
        + 
*/

#include "ros/ros.h"
#include "ros/console.h"

#include "geometry_msgs/Twist.h"

#include "message_pkg/Driver_respond.h"

#include <bits/stdc++.h>
using namespace std;

class filterTopic{
    public:
    float pub_frequence;
    string topicSub_driver1;
    string topicSub_driver2;
    string topicPub_driver1;
    string topicPub_driver2;

    // -- SUB
    ros::Subscriber sub_topic_driver1;
    message_pkg::Driver_respond driver1_respond;

    ros::Subscriber sub_topic_driver2;
    message_pkg::Driver_respond driver2_respond;

    // -- PUB
    ros::Publisher pub_filter1;
    message_pkg::Driver_respond driver1_filter;

    ros::Publisher pub_filter2;
    message_pkg::Driver_respond driver2_filter;

    // --
    float time_pub; // s
    double preTime_pub1;
    double preTime_pub2;
    // -- 
    double nowTime_speed1;
    double nowTime_speed2;
    bool is_timeout1;
    bool is_timeout2;
    float timeout;

	void driver1Respond_callback(const message_pkg::Driver_respond data){
		driver1_respond = data;
        // cout << "Tao co nhan duo du lieu tu driver 1" << endl;
		nowTime_speed1 = ros::Time::now().toSec();
    }

	void driver2Respond_callback(const message_pkg::Driver_respond data){
        // cout << "Tao co nhan duo du lieu tu driver 2" << endl;
		driver2_respond = data;
		nowTime_speed2 = ros::Time::now().toSec();
    }
};

int main(int argc, char **argv)
{
    cout << "Program start!";

    ros::init(argc, argv, "filterTopic_cpp");
    ros::NodeHandle n;

    filterTopic self;

    // -- parameter
    ros::param::get("frequence", self.pub_frequence);
    ros::Rate loop_rate(self.pub_frequence);

    ros::param::get("topicSub_driver1", self.topicSub_driver1);
    ros::param::get("topicSub_driver2", self.topicSub_driver2);

    ros::param::get("topicPub_driver1", self.topicPub_driver1);
    ros::param::get("topicPub_driver2", self.topicPub_driver2);

    // // -- SUB
    self.sub_topic_driver1 = n.subscribe(self.topicSub_driver1, 20, &filterTopic::driver1Respond_callback, &self);
    self.sub_topic_driver2 = n.subscribe(self.topicSub_driver2, 20, &filterTopic::driver2Respond_callback, &self);

    // -- PUB
    self.pub_filter1 = n.advertise<message_pkg::Driver_respond>(self.topicPub_driver1, 50);
    self.pub_filter2 = n.advertise<message_pkg::Driver_respond>(self.topicPub_driver2, 50);      

    // --
    self.time_pub = 1/self.pub_frequence; // s
    self.preTime_pub1 = ros::Time::now().toSec();
    self.preTime_pub2 = ros::Time::now().toSec();
    // -- 
    self.nowTime_speed1 = ros::Time::now().toSec();
    self.nowTime_speed2 = ros::Time::now().toSec();
    self.is_timeout1 = 0;
    self.is_timeout2 = 0;
    self.timeout = self.time_pub*3.;

    while(ros::ok()){

        double t1_out = ros::Time::now().toSec() - self.nowTime_speed1;
        if (t1_out > self.timeout){
            self.is_timeout1 = 1;
        }
        else{
            self.is_timeout1 = 0;
        }
        // --
        double t2_out = ros::Time::now().toSec() - self.nowTime_speed2;
        if (t2_out > self.timeout){
            self.is_timeout2 = 1;
        }
        else{
            self.is_timeout2 = 0;
        }
        // --
        if (self.is_timeout1 == 0){
            // cout<< "Im here!--------------------- pub ra driver1 ------------------" << endl;
            self.pub_filter1.publish(self.driver1_respond);
        }
        // --
        if (self.is_timeout2 == 0){
            // cout<< "Im here!--------------------- pub ra driver2 ------------------" << endl;
            self.pub_filter2.publish(self.driver2_respond);
        }        
        ros::spinOnce();     // allow receiving callbacks function
        loop_rate.sleep();
    }

    return 0;
}